Frameworks/CoreGraphics/DWriteWrapper.mm

```diff
@@ -479,4 +479,164 @@ HRESULT _DWriteFontGetBoundingBoxesForGlyphs(
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE CreateStreamFromKey(_In_ const void* fontFileReferenceKey,
+                                                  uint32_t fontFileReferenceKeySize,
+                                                  _Out_ IDWriteFontFileStream** fontFileStream) {
+        if (!fontFileStream) {
+            return E_INVALIDARG;
+        }
+
+        // Ignore first two params, just return the same kind of stream always
+        ComPtr<DWriteFontBinaryDataStream> ret;
+        RETURN_IF_FAILED(MakeAndInitialize<DWriteFontBinaryDataStream>(&ret, _dataProvider.get()));
+
+        *fontFileStream = ret.Detach();
+        return S_OK;
+    }
+
+private:
+    woc::unique_cf<CGDataProviderRef> _dataProvider;
+};
+
+/**
+ * Creates an IDWriteFontFace by attempting to use a CGDataProviderRef as a font file
+ */
+HRESULT _DWriteCreateFontFaceWithDataProvider(CGDataProviderRef dataProvider, IDWriteFontFace** outFontFace) {
+    ComPtr<IDWriteFactory> dwriteFactory;
+    RETURN_IF_FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), &dwriteFactory));
+
+    ComPtr<IDWriteFontFileLoader> dwriteDataLoader;
+    RETURN_IF_FAILED(MakeAndInitialize<DWriteFontBinaryDataLoader>(&dwriteDataLoader, dataProvider));
+
+    RETURN_IF_FAILED(dwriteFactory->RegisterFontFileLoader(dwriteDataLoader.Get()));
+
+    ComPtr<IDWriteFontFile> fontFile;
+
+    // First two params are not used (see DWriteFontBinaryDataLoader::CreateStreamFromKey())
+    // However, CreateCustomFontFileReference throws E_INVALIDARG if null is passed for the first parameter
+    // Pass in a pointer to an arbitrary, unused key, to get around this.
+    int unusedKey;
+    RETURN_IF_FAILED(dwriteFactory->CreateCustomFontFileReference(&unusedKey, sizeof(unusedKey), dwriteDataLoader.Get(), &fontFile));
+
+    // Analyze the font file for creating a font face
+    BOOL isSupportedFontType;
+    DWRITE_FONT_FILE_TYPE fontFileType;
+    DWRITE_FONT_FACE_TYPE fontFaceType;
+    uint32_t numberOfFaces;
+    RETURN_IF_FAILED(fontFile->Analyze(&isSupportedFontType, &fontFileType, &fontFaceType, &numberOfFaces));
+
+    RETURN_HR_IF(E_INVALIDARG, !isSupportedFontType);
+    RETURN_HR_IF(E_INVALIDARG, numberOfFaces == 0);
+
+    IDWriteFontFile* fontFileArray[] = { fontFile.Get() };
+
+    HRESULT created = dwriteFactory->CreateFontFace(fontFaceType,
+                                                    1, // number of elements in fontFileArray
+                                                    fontFileArray,
+                                                    0, // Just use the first face
+                                                    DWRITE_FONT_SIMULATIONS_NONE,
+                                                    outFontFace);
+
+    // Always try to unregister the font file loader, even if creation failed above
+    RETURN_IF_FAILED(dwriteFactory->UnregisterFontFileLoader(dwriteDataLoader.Get()));
```

Comment:
I guess I sort of see your point

---
In reply to: [88108767](https://github.com/Microsoft/WinObjC/pull/1386#discussion_r88108767) [](ancestors = 88108767)